service/service.go

```diff
@@ -100,36 +100,30 @@ func New(config Config) (*Service, error) {
 		GroupsClaim:   config.Viper.GetString(config.Flag.Service.Installation.Tenant.Kubernetes.API.Auth.Provider.OIDC.GroupsClaim),
 	}
 
-	var restConfig *rest.Config
-	{
-		c := k8srestconfig.Config{
-			Logger: config.Logger,
-
-			Address:    config.Viper.GetString(config.Flag.Service.Kubernetes.Address),
-			InCluster:  config.Viper.GetBool(config.Flag.Service.Kubernetes.InCluster),
-			KubeConfig: config.Viper.GetString(config.Flag.Service.Kubernetes.KubeConfig),
-			TLS: k8srestconfig.ConfigTLS{
-				CAFile:  config.Viper.GetString(config.Flag.Service.Kubernetes.TLS.CAFile),
-				CrtFile: config.Viper.GetString(config.Flag.Service.Kubernetes.TLS.CrtFile),
-				KeyFile: config.Viper.GetString(config.Flag.Service.Kubernetes.TLS.KeyFile),
-			},
-		}
-
-		restConfig, err = k8srestconfig.New(c)
-		if err != nil {
-			return nil, microerror.Mask(err)
-		}
-	}
-
 	var k8sClient *k8sclient.Clients
 	{
 		c := k8sclient.ClientsConfig{
 			SchemeBuilder: k8sclient.SchemeBuilder{
 				v1alpha1.AddToScheme,
 			},
-			Logger:     config.Logger,
-			RestConfig: restConfig,
+			Logger: config.Logger,
+		}
+
+		// Prefer KubeConfigPath if given, otherwise build restconfig.
+		{
+			inCluster := config.Viper.GetBool(config.Flag.Service.Kubernetes.InCluster)
+			kubeConfigPath := config.Viper.GetString(config.Flag.Service.Kubernetes.KubeConfigPath)
+			if !inCluster && kubeConfigPath != "" {
+				c.KubeConfigPath = kubeConfigPath
```

Comment:
I'd rather fail when `inCluster && kubeConfigPath != ""`

Reply:
Both are needed. One for driving logic and other one for validation.

Reply:
Why would you need `kubeConfigPath` if `inCluster` is true?

Reply:
No, you don't. By "both" I meant that both cases of logic are needed plus the `else` case for rest config. The validation is in the merged code.

Reply:
This is going wrong way :) Let me elaborate on my initial thought.

> I'd rather fail when `inCluster && kubeConfigPath != ""`

I mean if the user provides both `--incluster=true` and `--kubeconfigpath="non-empty"` this may be a misconfiguration / done by mistake. Someone may spin circles figuring why `--kubeconfigpath` flag isn't working not noticing that `--incluster=true` was specified. To prevent such situation I'd like to force the user to specify one or another by returning error on `inCluster && kubeConfigPath != ""` condition.

Reply:
Isn't that what my final code does?
`master`: https://github.com/giantswarm/azure-operator/blob/master/service/service.go#L116-L126

```go
			if !inCluster && kubeConfigPath != "" {
				c.KubeConfigPath = kubeConfigPath
			} else if inCluster && kubeConfigPath != "" {
				return nil, microerror.Maskf(invalidConfigError, "inCluster and kubeConfigPath must not be defined at the same time")
			} else {
				restConfig, err := buildK8sRestConfig(config)
				if err != nil {
					return nil, microerror.Mask(err)
				}
				c.RestConfig = restConfig
			}
```

Reply:
Ah cool sorry. I didn't notice that. This is a bit misleading. I would assume this should be the first check because then you don't have to check `!inCluster && kubeConfigPath != ""`. `kubeConfigPath != ""` is sufficient. But behaviour-wise this is good. Thanks!